File: GameState.java
//package perfect_connect_four;

import java.util.Arrays;
/*
bitstring representation:
board:
(6, 13, 20, 27, 34, 41, 48) <-- sentinel row, always filled with 0s
[5, 12, 19, 26, 33, 40, 47] << all rows below this represent the actual board
[4, 11, 18, 25, 32, 39, 46] << numbers in these rows only represent the bit count
[3, 10, 17, 24, 31, 38, 45]
[2, 9, 16, 23, 30, 37, 44]
[1, 8, 15, 22, 29, 36, 43]
[0, 7, 14, 21, 28, 35, 42]

note that this class is based on information gained in
https://towardsdatascience.com/creating-the-perfect-connect-four-ai-bot-c165115557b0
*/
public class GameState {
    private int[] bitRepCur; //position of current player's pieces
    private int[] bitRepMask; //all pieces
    private int[] bitRepOpp; //opponent pieces
    private long bitstringCur = 0;
    private long bitstringMask = 0;
    private long bitstringOpp = 0;
    public int totalMoves = 0;
    public final int WIDTH = 7;
    public final int HEIGHT = 6;
    public final int MINIMUM_SCORE = -(WIDTH * HEIGHT + 1)/2 - 3;
    public final int MAXIMUM_SCORE = (WIDTH * HEIGHT + 1)/2 - 3;
    public int perspective = 0; // if 1, the current player is the user, 2 is the computer
    public String[] perspText = new String[]{"", "User", "AI"};

    /*public GameState(int[][] board) {
        int i = 0;
        int bits = (board.length) * board[0].length;
        bitRepComp = new int[bits];
        bitRepMask = new int[bits];
        bitRepPl = new int[bits];
        for(int col = board[0].length - 1; col >= 0; col--){
            for(int row = 0; row < board.length; row++) {
                if(board[row][col] == 2) bitRepComp[i] = 1;
                if(board[row][col] != 0) bitRepMask[i] = 1;
                bitRepPl[i] = bitRepComp[i]^bitRepMask[i];
                i++;
                int power = bits - i;
                bitstringComp += Math.pow(2, power) * bitRepComp[i-1];
                bitstringMask += Math.pow(2, power) * bitRepMask[i-1];
                bitstringPl += Math.pow(2, power) * bitRepPl[i-1];
            }
        }
        //System.out.println(bitstringComp + " " + bitstringMask + " " + bitstringPl);
        //System.out.println(Arrays.toString(bitRepMask));
    }*/

    public GameState(int perspective, int totalMoves) {
        this.perspective = perspective;
        this.totalMoves = totalMoves;
    }

    public GameState(long bitstringCur, long bitstringOpp, long mask, int perspective, int totalMoves) { //bitstring takes for the object being made in reference to the current player
        this.bitstringCur = bitstringCur;
        this.bitstringOpp = bitstringOpp;
        this.bitstringMask = mask;
        this.perspective = perspective;
        this.totalMoves = totalMoves;
        //System.out.println("AI: " + bitstringComp + " Player: " + bitstringPl + " Mask: " + bitstringMask);
    }

    public boolean winner() {
        long m = 0;
        //horizontal check
        m = bitstringCur & (bitstringCur >> 7);
        if((m & (m >> 14)) > 0)
            return true;
        //diagonal \ check
        m = bitstringCur & (bitstringCur >> 6);
        if((m & (m >> 12)) > 0)
            return true;
        //diagonal / check
        m = bitstringCur & (bitstringCur >> 8);
        if((m & (m >> 16)) > 0)
            return true;
        //vertical check
        m = bitstringCur & (bitstringCur >> 1);
        if((m & (m >> 2)) > 0)
            return true;
        //no win pattern found
        return false;
    }

    public boolean loser() {
        long m = 0;
        //horizontal check
        m = bitstringOpp & (bitstringOpp >> 7);
        if((m & (m >> 14)) > 0)
            return true;
        //diagonal \ check
        m = bitstringOpp & (bitstringOpp >> 6);
        if((m & (m >> 12)) > 0)
            return true;
        //diagonal / check
        m = bitstringOpp & (bitstringOpp >> 8);
        if((m & (m >> 16)) > 0)
            return true;
        //vertical check
        m = bitstringOpp & (bitstringOpp >> 1);
        if((m & (m >> 2)) > 0)
            return true;
        //no win pattern found
        return false;
    }

    public long key(){
        return bitstringCur + bitstringMask;
    }

    public GameState move(int col) { //the current player moves
        //piece = 1: player, = 2: computer
        long nextOppPos = bitstringCur ^ bitstringMask; //gives opponent position
        long nextMask = bitstringMask | (bitstringMask + (1L << (long)(col * 7))); //new mask
        long nextPos = nextOppPos ^ nextMask; //current player's position
        int newMovesNum = totalMoves + 1;
        GameState newState = new GameState(nextOppPos, nextPos, nextMask, (perspective % 2) + 1, newMovesNum); //this state is now in a different perspective (swapped to simulate turns)
        return newState;
        /*if(piece == 1) {
            long nextPosComp = bitstringPl ^ bitstringMask;
            long nextMask = bitstringMask | (bitstringMask + (1L << (long)(col * 7)));
            GameState newState = new GameState(nextPosComp, nextMask, 2);
            return newState;
        }
        else if (piece == 2) {
            long nextPosPl = bitstringComp ^ bitstringMask;
            long nextMask = bitstringMask | (bitstringMask + (1L << (long)(col * 7)));
            GameState newState = new GameState(nextPosPl, nextMask, 1);
            return newState;
        }
        return null;*/
    }

    public String getInfo()
    {
        String currentInfo = perspText[perspective] + ": " + bitstringCur + ", " + perspText[(perspective % 2) + 1] + ": " + bitstringOpp + ", Mask: " + bitstringMask;
        return currentInfo;
    }

    public long[] getBitstrings() {
        long[] out = new long[]{bitstringCur, bitstringMask, bitstringOpp};
        return out;
    }

    public boolean canMove(int col) {
        long topPieceMask = (1L << (HEIGHT - 1)) << col*(HEIGHT+1);
        return (bitstringMask & topPieceMask) == 0;
    }

    

    
}

File: gamestate.cpp
#include <bitset>
#include "gamestate.h"

GameState::GameState() 
{
    curBitMap = 0;
    oppBitMap = 0;
    maskBitMap = 0;
    totalMoves = 0;   
}
GameState::GameState(long long curBitMap, long long oppBitMap, long long mask)
{
    (*this).curBitMap = curBitMap;
    (*this).oppBitMap = oppBitMap;
    (*this).maskBitMap = mask;
    totalMoves = 0;
}

long long GameState::sentinelRow()
{
    long long topRow = 283691315109952LL; //bitmap of entire sentinel row of 1s
    return topRow;
}

long long GameState::outsideBoard()
{
    long long outside = 9222809086901354496LL;
    return outside;
}

bool GameState::canMove(int col)
{
    long long place = (1LL << (HEIGHT - 1)) << col*(HEIGHT + 1);
    long long m = maskBitMap & place;
    if(m == 0) return true;
    else return false;
}
long long GameState::key()
{
    return curBitMap + maskBitMap;
}

GameState GameState::move(int col)
{
    long long nextMask = maskBitMap | (maskBitMap + (1LL << (long long)(col*7)));
    long long nextOpp = oppBitMap ^ nextMask;
    long long nextCur = nextOpp ^ nextMask;
    GameState newState(nextCur, nextOpp, nextMask);
    totalMoves++;
    return newState;
}

bool GameState::winner()
{
    long long map = 0;
    map = curBitMap & (curBitMap >> 7);
    if((map & (map >> 14)) > 0)
        return true;
    //diagonal \ check
    map = curBitMap & (curBitMap >> 6);
    if((map & (map >> 12)) > 0)
        return true;
    //diagonal / check
    map = curBitMap & (curBitMap >> 8);
    if((map & (map >> 16)) > 0)
        return true;
    //vertical check
    map = curBitMap & (curBitMap >> 1);
    if((map & (map >> 2)) > 0)
        return true;
    //no win pattern found
    return false;
}

bool GameState::loser()
{
    long long map = 0;
    map = oppBitMap & (oppBitMap >> 7);
    if((map & (map >> 14)) > 0)
        return true;
    //diagonal \ check
    map = oppBitMap & (oppBitMap >> 6);
    if((map & (map >> 12)) > 0)
        return true;
    //diagonal / check
    map = oppBitMap & (oppBitMap >> 8);
    if((map & (map >> 16)) > 0)
        return true;
    //vertical check
    map = oppBitMap & (oppBitMap >> 1);
    if((map & (map >> 2)) > 0)
        return true;
    //no win pattern found
    return false;
}

int GameState::bitCount(long long map)
{
    bitset<64> m(map);
    return m.count();
}

int GameState::centerPieces(long long map)
{
    long long centerColMask = ((1LL << HEIGHT) - 1LL) << (HEIGHT + 1)*3; //mask of the entire center column
    long long toCount = centerColMask & map;
    return bitCount(toCount);
}

int GameState::checkThrees(long long map)
{
    long long avail = ~maskBitMap;
    avail = avail - sentinelRow();
    int count = 0;
    long long m1 = 0;
    long long m2 = 0;
    //horizontal
    //shift tokens right
    m1 = map & (map << 7);
    m1 = ((m1 & (m1 << 7)) << 7) & avail;
    count += bitCount(m1);
    //shift tokens left
    m2 = map & (map >> 7);
    m2 = ((m2 & (m2 >> 7)) >> 7) & avail;
    count += bitCount(m2);

    //vertical
    //shift up
    m1 = map & (map << 1);
    m1 = ((m1 & (m1 << 1)) << 1) & avail;
    count += bitCount(m1);
    //shift down
    m2 = map & (map >> 1);
    m2 = ((m2 & (m2 >> 1)) >> 1) & avail;
    count += bitCount(m2);
    
    //diagonal \
    //shift tokens to the upper right
    m1 = map & (map << 6);
    m1 = ((m1 & (m1 << 6)) << 6) & avail;
    count += bitCount(m1);
    //shift tokens to the lower left
    m2 = map & (map >> 6);
    m2 = ((m2 & (m2 >> 6)) >> 6) & avail;
    count += bitCount(m2);
    
    //diagonal /
    //shift tokens to the lower right
    m1 = map & (map << 8);
    m1 = ((m1 & (m1 << 8)) << 8) & avail;
    count += bitCount(m1);
    //shift tokens to the upper left
    m2 = map & (map >> 8);
    m2 = ((m2 & (m2 >> 8)) >> 8) & avail;
    count += bitCount(m2);
    
    return count;
}

int GameState::checkTwos(long long map)
{
    int count = 0;
	long long avail = ~maskBitMap;
    avail = avail - sentinelRow() - outsideBoard();
	long long m1 = 0;
	long long m2 = 0;
	
	//horizontal
    //shift tokens right
    m1 = map & (map << 7);
    m1 = (((m1 << 7) & avail) << 7) & avail;
    count += bitCount(m1);
    //shift tokens left
    m2 = map & (map >> 7);
    m2 = (((m2 >> 7) & avail) >> 7) & avail;
    count += bitCount(m2);

    //vertical
    //shift up
    m1 = map & (map << 1);
    m1 = (((m1 << 1) & avail) << 1) & avail;
    count += bitCount(m1);
    //shift down
    m2 = map & (map >> 1);
    m2 = (((m2 >> 1) & avail) >> 1) & avail;
    count += bitCount(m2);
    
    //diagonal \
    //shift tokens to the upper right
    m1 = map & (map << 6);
    m1 = (((m1 << 6) & avail) << 6) & avail;
    count += bitCount(m1);
    //shift tokens to the lower left
    m2 = map & (map >> 6);
    m2 = (((m2 >> 6) & avail) >> 6) & avail;
    count += bitCount(m2);
    
    //diagonal /
    //shift tokens to the lower right
    m1 = map & (map << 8);
    m1 = (((m1 << 8) & avail) << 8) & avail;
    count += bitCount(m1);
    //shift tokens to the upper left
    m2 = map & (map >> 8);
    m2 = (((m2 >> 8) & avail) >> 8) & avail;
    count += bitCount(m2);
    
	return count;
}

int GameState::score()
{
    int s = 0;
	s += (centerPieces(curBitMap) * 30);
	s += (checkTwos(curBitMap) * 10);
	s += (checkThrees(curBitMap) * 80);
	if(winner()) s += 100000;
	
	//System.out.println("center: " + centerPieces());
	//System.out.println("2s: " + checkTwos());
	//System.out.println("3s: " + checkThrees());
	
	s -= (centerPieces(oppBitMap) * 30);
	s -= (checkTwos(oppBitMap) * 10);
	s -= (checkThrees(oppBitMap) * 80);
	if(loser()) s -= 100000;
	
	return s;
}

int GameState::getTotalMoves()
{
    return totalMoves;
}

long long GameState::getCurBM()
{
    return curBitMap;
} 

long long GameState::getOppBM()
{
    return oppBitMap;
}

long long GameState::getMaskBM()
{
    return maskBitMap;
}
